cilium/CFP-22215-mutual-auth-for-service-mesh.md

```diff
@@ -0,0 +1,442 @@
+
+## Goals
+
+* Provide a mechanism for pods to opt into peer authentication
+* Pluggable for existing certificate management systems (SPIFFE, Vault, SMI, Istio, cert-manager, …)
+* Configurable certificate granularity
+* Leverage existing encryption protocol support for the dataplane
+
+
+## Non-Goals
+
+* Inline authentication via a proxy
+* Drop less connection establishment
+
+
+## Proposal
+
+### Overview
+
+The Isovalent blog Next-Generation Mutual Authentication with Cilium Service Mesh described the high level design for mutual authentication where there are two channels for connectivity: Firstly, a control plane connection between cilium-agent instances on each node provides authentication of connections between pods on the nodes. Secondly, the existing Cilium encryption support using WireGuard or IPsec provides an encrypted dataplane for the connections. This document seeks to describe the per-node implementation that coordinates these out-of-band authentication channels with the in-band network encryption channels.
+
+Here is a diagram showing the key parts of the design
+
+![mTLS General Architecture](./images/mTLS-general-architecture.png)
+
+The following sub-sections are broken up by the aspects of the design where specific changes need to be made, starting from the k8s resource types that need to be modified, and followed by the considerations for the cert management control plane, the (local) control plane, and the data plane. While the following sections may focus more on SPIFFE as a default auth implementation for the pluggable interface, the same principles should apply for other auth mechanisms.
+
+### Resources
+
+TODO: This section needs an update.
+
+#### ConfigMap - cilium-config
+
+* option-name: `default`
+  * description
+* mesh-auth-mode: `disabled`
+  * enabled / disabled
+* mesh-auth-cert-source: `spiffe`
+  * Configure certificate mechanism, default to directory path
+* mesh-spire-server: `____`
+  * Configure the spire server
+* mesh-spire-agent: `unix://var/run/cilium/spiffe/agent.sock`
+  * Configure the SPIRE agent
+* mesh-auth-path: `/var/run/cilium/auth/`
+  * Configure the directory to pull certificates from, if mesh-auth-cert is directory
+* mesh-auth-port: `4250`
+  * Each agent opens this port for mutual authentication
+  * TODO review port for conflicts
+
+#### CRD - CiliumNetworkPolicy, CiliumClusterwideNetworkPolicy
+
+* Extend the existing policy structure to include a new Authentication field
+
+```yaml
+apiVersion: "cilium.io/v2"
+kind: CiliumNetworkPolicy
+metadata:
+  name: "l3-egress-rule"
+spec:
+  endpointSelector:
+    matchLabels:
+      run: ssh-client
+  egress:
+  - toEndpoints:
+    - matchLabels:
+        run: sshd
+    auth:
+       enabled: true
```

Comment:
This doesn't align with the current implementation where the `auth type` can be configured.

If it should only be an `enabled` flag, we might need to add the `auth type` in the config map as dedicated property.

Reply:
I didn't realize that's how our current impl worked, I'll update.